[GStreamer][MSE] Append pipeline is counting streams incorrectly with new init segments

calvaris opened the pull request:
#### 847e28d6e49b2c43727efab967296f483aec55eb
<pre>
[GStreamer][MSE] Append pipeline is counting streams incorrectly with new init segments
<a href="https://bugs.webkit.org/show_bug.cgi?id=254664">https://bugs.webkit.org/show_bug.cgi?id=254664</a>

Reviewed by Alicia Boya Garcia.

When you append another init segment we match GStreamer pads against tracks but we are not counting properly as we are
skipping the pads that are already linked we can end up with less pads than tracks. Besides it is against the spec. We
need to match the number of tracks and track IDs if there are more than one for a kind.

* LayoutTests/media/media-source/media-source-seek-detach-crash.html: Fail the test earlier instead of waiting for a
timeout.
* LayoutTests/platform/glib/TestExpectations: Unmark
media/encrypted-media/encrypted-media-append-encrypted-unencrypted.html as failure. It works now.
* Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
(WebCore::AppendPipeline::didReceiveInitializationSegment):
(WebCore::AppendPipeline::recycleTrackForPad):

Canonical link: <a href="https://commits.webkit.org/264676@main">https://commits.webkit.org/264676@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b483c022009e8897e40a71571091d37375fc76ac

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/8339 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/8632 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/8849 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/10007 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/8404 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/8348 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/10620 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/8543 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/11271 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/8485 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/9541 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/7561 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/10152 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/6839 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/7630 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/15191 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/7960 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/7773 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/11118 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/8244 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/6718 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/7533 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/7542 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2002 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/11743 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/7987 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
```diff
@@ -459,36 +459,56 @@ void AppendPipeline::didReceiveInitializationSegment()
         }
 
-        if (videoPadsCount < videoTracksCount || audioPadsCount < audioTracksCount || textPadsCount < textTracksCount) {
+        // Step 3 of the MSE spec append initialization segment algorithm.
```
Please quote the spec here word by word instead of just referencing it by step number, as that is less clear and can become very confusing if the spec is later edited. For instance:

```
// 4.5.7.3.1: Verify the following properties. If any of the checks fail then run the append error algorithm and abort these steps.
// 4.5.7.3.1.1: The number of audio, video, and text tracks match what was in the first initialization segment.
bool doTracksMatch = audioPadStreamIDs.size() == audioTracksCount && videoPadStreamIDs.size() == videoTracksCount && textPadStreamIDs.size() == textTracksCount;
// 4.5.7.3.1.2: If more than one track for a single type are present (e.g., 2 audio tracks), then the Track IDs match the ones in the first initialization segment.
if (doTracksMatch && audioTracksCount > 1)
    doTracksMatch = doAudioTrackStreamIDsMatch;
if (doTracksMatch && videoTracksCount > 1)
    doTracksMatch = doVideoTrackStreamIDsMatch;
if (doTracksMatch && textTracksCount > 1)
    doTracksMatch = doTextTrackStreamIDsMatch;
```

calvaris replied:
Solved.

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
```diff
@@ -459,36 +459,56 @@ void AppendPipeline::didReceiveInitializationSegment()
+            if (track->streamType == StreamType::Audio) {
                 audioTracksCount++;
-            else if (track->streamType == StreamType::Video)
+                if (streamID.isEmpty() || !audioPadStreamIDs.contains(streamID))
```
I would like to know when is the case that `streamID` is empty.

calvaris replied:
Me too.

calvaris replied:
TL&DR

If I added it is because it was in layout test that crashed because of that but I lost track of all regressions I fixed while iterating on this patch.

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
```diff
@@ -837,7 +857,8 @@ bool AppendPipeline::recycleTrackForPad(GstPad* demuxerSrcPad)
     }
 
-    if (!matchingTrack->isLinked())
+    GRefPtr<GstCaps> matchingTrackCaps = adoptGRef(gst_pad_get_current_caps(matchingTrack->entryPad.get()));
+    if (!matchingTrack->isLinked() && (!matchingTrackCaps || gst_caps_can_intersect(parsedCaps.get(), matchingTrackCaps.get())))
```
Good catch.

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
```diff
@@ -459,36 +459,56 @@ void AppendPipeline::didReceiveInitializationSegment()
+            UNUSED_VARIABLE(parsedCaps);
+            UNUSED_VARIABLE(presentationSize);
+            GUniquePtr<char> streamIDAsCharacters(gst_pad_get_stream_id(pad));
+            String streamID = String::fromLatin1(streamIDAsCharacters.get());
```
You don't need two variables.

    String streamID = String::fromLatin1(GUniquePtr<char>(gst_pad_get_stream_id(pad)).get());

> All temporary objects are destroyed as the last step in evaluating the [full-expression](https://en.cppreference.com/w/cpp/language/expressions#Full-expressions) that (lexically) contains the point where they were created, and if multiple temporary objects were created, they are destroyed in the order opposite to the order of creation. This is true even if that evaluation ends in throwing an exception.

calvaris replied:
Solved

ntrrgc reviewed (approved):
Thanks for this patch.

`gst_pad_get_stream_id()` is a great find. I have been able to confirm through looking at the source code that both qtdemux and matroskademux emit the track IDs found in the containers metadata (namely the `tkhd` box for MP4 and the [TrackNumber](https://www.matroska.org/technical/elements.html#TrackNumber):[TrackUID](https://www.matroska.org/technical/elements.html#TrackUID) pair for Matroska/WebM), and this is not a new feature, already being present in GStreamer 1.10 and earlier.

So far, for lack of a way to get these bits of information, AppendPipeline has been using pad indices as track IDs (see `AppendPipeline::generateTrackId()`), so the first video pad gets V1, the second V2 and so on...

Now, taking this into consideration, it makes sense to update the code and replace the autogenerated track IDs with actual track IDs from GStreamer. I also noticed that the MediaSource spec requires track IDs exposed to the user to be distinct but doesn't specify they need to match the exact values in tkhd (although the engine internally must match those).

Something I've been wondering about this patch is, how are we handling recycling pads? This is the current code for matching tracks and pads:

    // Try to find a matching pre-existing track. Ideally, tracks should be matched by track ID, but matching by type
    // is provided as a fallback -- which will be used, since we don't have a way to fetch those from GStreamer at the moment.
    Track* matchingTrack = nullptr;
    for (std::unique_ptr<Track>& track : m_tracks) {
        if (track->streamType != streamType)
            continue;
        matchingTrack = &*track;
        if (track->trackId == trackId)
            break;
    }

Assuming pad indices are consistent, our code should still work. For that reason, this patch LGTM as it is. Feel free to rework the track IDs to use the data from GStreamer stream IDs if you want, but that can be a future patch as well, leaving this fix contained.

calvaris committed:
[GStreamer][MSE] Append pipeline is counting streams incorrectly with new init segments
https://bugs.webkit.org/show_bug.cgi?id=254664

Reviewed by Alicia Boya Garcia.

When you append another init segment we match GStreamer pads against tracks but we are not counting properly as we are
skipping the pads that are already linked we can end up with less pads than tracks. Besides it is against the spec. We
need to match the number of tracks and track IDs if there are more than one for a kind.

* LayoutTests/media/media-source/media-source-seek-detach-crash.html: Fail the test earlier instead of waiting for a
timeout.
* LayoutTests/platform/glib/TestExpectations: Unmark
media/encrypted-media/encrypted-media-append-encrypted-unencrypted.html as failure. It works now.
* Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
(WebCore::AppendPipeline::didReceiveInitializationSegment):
(WebCore::AppendPipeline::recycleTrackForPad):

Canonical link: https://commits.webkit.org/264676@main
